### PR TITLE
Use file volume mount for memberlist secret key

### PR DESCRIPTION
**Pull request by liornoy**
This PR changes the way we insert the memberlist secret key to the speaker.
Switching from env var to a file in volume mount.

### CODE REVIEW

**Comment by liornoy** [comment]
e2e-use-operator fails because the operator deploys metallb using YAMLs that inject env var instead of mounting a file.
so it's a chicken-egg situation, where we'll have to merge this and then right after sending a fix to the operator.

File: charts/metallb/templates/speaker.yaml [hunk 1]
```diff
@@ -288,11 +292,15 @@ spec:
             - ALL
             add:
             - NET_RAW
-        {{- if .Values.speaker.frr.enabled }}
         volumeMounts:
+          {{- if .Values.speaker.memberlist.enabled }}
```
Comment (fedepaol) [inline charts/metallb/templates/speaker.yaml:296]: what if neither memberlist or frr are enabled?

File: charts/metallb/templates/speaker.yaml [hunk 2]
```diff
@@ -147,6 +147,15 @@ spec:
...
       hostNetwork: true
       volumes:
+      {{- if .Values.speaker.memberlist.enabled }}
+        - name: memberlist
+          secret:
+            secretName: {{ include "metallb.secretName" . }}
+            defaultMode: 420
+            items:
+            - key: secretkey
+              path: metallb/speaker
```
Comment (fedepaol) [inline charts/metallb/templates/speaker.yaml:157]: just "key" as the rest is given by the full path

File: speaker/main.go [hunk 3]
```diff
@@ -117,7 +116,13 @@ func main() {
 	}()
 	defer level.Info(logger).Log("op", "shutdown", "msg", "done")
 
-	sList, err := speakerlist.New(logger, *myNode, *mlBindAddr, *mlBindPort, *mlSecret, *namespace, *mlLabels, stopCh)
+	mlSecret, err := os.ReadFile("/etc/ml_secret_key/metallb/speaker")
```
Comment (fedepaol) [inline speaker/main.go:119]: Let's pass the path as a parameter

File: speaker/main.go [hunk 4]
```diff
@@ -117,7 +116,13 @@ func main() {
...
+	if err != nil {
```
Comment (fedepaol) [inline speaker/main.go:120]: The secret is optional, so it's fine not to have it. By doing this we are making it mandatory

File: charts/metallb/templates/speaker.yaml (hunk 1 above)
Comment (liornoy) [reply charts/metallb/templates/speaker.yaml:296]: done. added an OR statement above.

File: charts/metallb/templates/speaker.yaml (hunk 2 above)
Comment (liornoy) [reply charts/metallb/templates/speaker.yaml:157]: done

File: speaker/main.go (hunk 3 above)
Comment (liornoy) [reply speaker/main.go:119]: done

File: speaker/main.go (hunk 4 above)
Comment (liornoy) [reply speaker/main.go:120]: done

**Comment by liornoy** [comment]
(Updated speaker main.go - to include the const mlSecretKeyName = "secretkey"

File: charts/metallb/policy/speaker.rego [hunk 5]
```diff
@@ -11,15 +11,14 @@ deny[msg] {
...
+	not input.spec.template.spec.containers[0].env[5].name == "METALLB_ML_SECRET_KEY_PATH"
+	msg = "speaker env does not contain METALLB_ML_SECRET_KEY_PATH at env[5]"
 }
 
 deny[msg] {
 	input.kind == "DaemonSet"
-	not input.spec.template.spec.containers[0].env[5].valueFrom.secretKeyRef.name == "RELEASE-NAME-metallb-memberlist"
-	not input.spec.template.spec.containers[0].env[5].valueFrom.secretKeyRef.key == "secretkey"
-	msg = "speaker env METALLB_ML_SECRET_KEY secretKeyRef does not equal expected value"
+	not input.spec.template.spec.containers[0].env[5].value == "/etc/ml_secret_key"
```
Comment (fedepaol) [inline charts/metallb/policy/speaker.rego:20]: The value can be overridden, this is static, so if the user changes it this check will always fail.

File: speaker/main.go [hunk 6]
```diff
@@ -61,13 +62,15 @@ type service interface {
 func main() {
 	prometheus.MustRegister(announcing)
 
+	const mlSecretKeyName = "secretkey"
```
Comment (fedepaol) [inline speaker/main.go:65]: Having this as a constant while it's used only in one place is not adding value. On the other hand, I think it makes sense to have it as a global constant in k8s.go and use it also there (where we create the secret).

File: charts/metallb/policy/speaker.rego (hunk 5 above)
Comment (liornoy) [reply charts/metallb/policy/speaker.rego]: you're right, removed.

File: speaker/main.go (hunk 6 above)
Comment (liornoy) [reply speaker/main.go]: done! thanks

**Comment by fedepaol** [comment]
LGTM, thanks!

**Pull request merged**